Enable TensorPipe CUDA sending to self

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50676 Let TensorPipe detect peer access
* #50675 Enable TensorPipe CUDA fallback channel
* **#50674 Enable TensorPipe CUDA sending to self**

Differential Revision: [D25941964](https://our.internmc.facebook.com/intern/diff/D25941964)

mrshenli committed:
Enable TensorPipe CUDA sending to self

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -85,6 +85,7 @@ constexpr int64_t kBasicChannelPriority = 0;
 
 #if TENSORPIPE_HAS_CUDA_IPC_CHANNEL && defined(USE_CUDA_NOT_ROCM)
 constexpr int64_t kCudaIpcChannelPriority = 300;
+constexpr int64_t kCudaXthChannelPriority = 400;
```
This shouldn't be guarded by the `TENSORPIPE_HAS_CUDA_IPC_CHANNEL`. That macro only control the CUDA_IPC channel.

mrshenli replied:
ah, yep, good catch!

mrshenli committed:
Update on "Enable TensorPipe CUDA sending to self"


Differential Revision: [D25941964](https://our.internmc.facebook.com/intern/diff/D25941964)

[ghstack-poisoned]

mrshenli committed:
Update on "Enable TensorPipe CUDA sending to self"


Differential Revision: [D25941964](https://our.internmc.facebook.com/intern/diff/D25941964)

[ghstack-poisoned]
